blockchain/timestamping.py

```diff
@@ -0,0 +1,211 @@
+        Args:
+            network: mainnet, regtest, testnet
+            fee_provider: suggests fee per bytes to pay
+        """
+        bitcoin.SelectParams(network)
+        self.fee_provider = fee_provider
+
+
+    def ispersisted(self, tx_id, data):
+        """
+        Verifies ``data`` item is on the bitcoin transaction identified by the ``tx_id``
+        Args:
+            tx_id: the bitcoin transaction id where the data is stored
+            data: the item data we want to verify
+
+        Returns:
+            True if the ``data`` item is in the bitcoin transaction identified by ``tx_id``; otherwise False
+        """
+        log.info('Connection to local bitcoin node')
+        proxy = bitcoin.rpc.Proxy()
+        tx = proxy.getrawtransaction(tx_id, True)['tx']
+        for vout in tx.vout:
+            step = 'init'
+            for op in vout.scriptPubKey:
+                if step == 'init' and op == OP_RETURN:
+                    step = 'waiting_data'
+                    continue
+                if step == 'waiting_data' and op == data:
+                    return True
+        return False
+
+
+    def persist(self, data):
+        """
+        Persists data in the Bitcoin blockchain and returns and transaction id for future reference.
+
+        Connects to the local bitcoin node using the bitcoin.conf file RPC username and password,
+        requests the list of unspent transaction outputs (UTXO) and takes the one with less amount
+        from being used as input in the new transaction.
+
+
+        Args:
+            data: the item data to be persisted
+
+        Returns:
+            A transaction id for referencing the saved data item in the transaction output
+        """
+        proxy = bitcoin.rpc.Proxy()
+        utxo = sorted(proxy.listunspent(0), key=lambda x: hash(x['amount']))[-1]
+        tx = self._build_transaction(utxo, self.get_new_pubkey(), data)
+        value_in = utxo['amount']
+
+        fee_byte = self.fee_provider.recommended()
+        fee_byte = max(fee_byte, self.MIN_FEE_BYTE)
+        assert fee_byte < self.MAX_FEE_BYTE, "too high fee"
+
+        new_tx = tx
+        suggested_fee = (len(new_tx.serialize()) + self.P2PKH_SIGSCRIPT_SIZE) * fee_byte
+        new_tx.vout[0].nValue = int(value_in - suggested_fee)
+        r = proxy.signrawtransaction(new_tx)
+        assert r['complete']
+        new_tx = r['tx']
+        suggested_fee = len(new_tx.serialize()) * fee_byte
+        tx_id = proxy.sendrawtransaction(new_tx)
+        log.info('Transaction sent')
+        return tx_id
+
+
+    def get_new_pubkey(self):
+        """
+        Requests a new bitcoin address and returns its public key
+        """
+        proxy = bitcoin.rpc.Proxy()
+        address = proxy.getnewaddress()
+        pubkey = proxy.validateaddress(address)['pubkey']
+        return pubkey
+
+
+    def _build_transaction(self, output, change_pubkey, data):
```

Comment:
Refactored the function call. Do you remember what the length allowable with OP_RETURN?